### ISSUE TITLE
Finish reason: 9. Finish message: Malformed function call:

### ISSUE THREAD

**Issue by b0noI** [googleapis/python-aiplatform#4472]
#### Environment details

  - OS type and version: Mac OS 14.5
  - Python version: Python 3.12.5
  - pip version: `pip 24.2`
  - `google-cloud-aiplatform` version: `Version: 1.68.0

#### Steps to reproduce

git clone git@github.com:GeminiAgentsToolkit/gemini-agents-toolkit.git
pip install -r ./requirements.txt
python ./examples/investing_pipeline_example.py

run the code several times, on some day it works almost all the time, today it was working 2 out of 10 times only T_T

output example:

```(toolbox) ➜  gemini-toolbox git:(main) ✗ python ./examples/investing_pipeline_example.py
Retrying send_message... Attempt #1, Last error: The model response did not complete successfully.
Finish reason: 9.
Finish message: Malformed function call:
print(28 > 30)
.
Safety ratings: [].
To protect the integrity of the chat session, the request and response were not added to chat history.
To skip the response validation, specify `model.start_chat(response_validation=False)`.
Note that letting blocked or otherwise incomplete responses into chat history might lead to future interactions being blocked by the service.
Retrying send_message... Attempt #1, Last error: The model response did not complete successfully.
Finish reason: 9.
Finish message: Malformed function call:
price = 121.3 * 1.04
print(default_api.set_limit_sell_order(price=price))
.
Safety ratings: [].
To protect the integrity of the chat session, the request and response were not added to chat history.
To skip the response validation, specify `model.start_chat(response_validation=False)`.
Note that letting blocked or otherwise incomplete responses into chat history might lead to future interactions being blocked by the service.
Retrying send_message... Attempt #2, Last error: The model response did not complete successfully.
Finish reason: 9.
Finish message: Malformed function call:
print(157.3 * 1.04)
.
Safety ratings: [].
To protect the integrity of the chat session, the request and response were not added to chat history.
To skip the response validation, specify `model.start_chat(response_validation=False)`.
Note that letting blocked or otherwise incomplete responses into chat history might lead to future interactions being blocked by the service.
Traceback (most recent call last):
  File "/Users/vkovalevskyi/miniforge3/envs/toolbox/lib/python3.12/site-packages/tenacity/__init__.py", line 478, in __call__
    result = fn(*args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^
  File "/Users/vkovalevskyi/src/gemini-toolbox/gemini_agents_toolkit/agent.py", line 201, in send_message
    response = self.chat.send_message(
               ^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/vkovalevskyi/miniforge3/envs/toolbox/lib/python3.12/site-packages/vertexai/generative_models/_generative_models.py", line 1165, in send_message
    return self._send_message(
           ^^^^^^^^^^^^^^^^^^^
  File "/Users/vkovalevskyi/miniforge3/envs/toolbox/lib/python3.12/site-packages/vertexai/generative_models/_generative_models.py", line 1293, in _send_message
    self._response_validator(
  File "/Users/vkovalevskyi/miniforge3/envs/toolbox/lib/python3.12/site-packages/vertexai/generative_models/_generative_models.py", line 1070, in _validate_response
    raise ResponseValidationError(
vertexai.generative_models._generative_models.ResponseValidationError: The model response did not complete successfully.
Finish reason: 9.
Finish message: Malformed function call:
print(157.3 * 1.04)
.
Safety ratings: [].
To protect the integrity of the chat session, the request and response were not added to chat history.
To skip the response validation, specify `model.start_chat(response_validation=False)`.
Note that letting blocked or otherwise incomplete responses into chat history might lead to future interactions being blocked by the service.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/vkovalevskyi/src/gemini-toolbox/./examples/investing_pipeline_example.py", line 103, in <module>
    pipeline.step("set limit sell order for TQQQ for price +4% of current price", history=history)
  File "/Users/vkovalevskyi/src/gemini-toolbox/gemini_agents_toolkit/pipeline/eager_pipeline.py", line 55, in step
    self.prev_step_data = agent_to_use.send_message(prompt)
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/vkovalevskyi/miniforge3/envs/toolbox/lib/python3.12/site-packages/tenacity/__init__.py", line 336, in wrapped_f
    return copy(f, *args, **kw)
           ^^^^^^^^^^^^^^^^^^^^
  File "/Users/vkovalevskyi/miniforge3/envs/toolbox/lib/python3.12/site-packages/tenacity/__init__.py", line 475, in __call__
    do = self.iter(retry_state=retry_state)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/vkovalevskyi/miniforge3/envs/toolbox/lib/python3.12/site-packages/tenacity/__init__.py", line 376, in iter
    result = action(retry_state)
             ^^^^^^^^^^^^^^^^^^^
  File "/Users/vkovalevskyi/miniforge3/envs/toolbox/lib/python3.12/site-packages/tenacity/__init__.py", line 419, in exc_check
    raise retry_exc from fut.exception()
tenacity.RetryError: RetryError[<Future at 0x10c6313d0 state=finished raised ResponseValidationError>]
```

**Comment by b0noI** [comment]
any updates?

**Comment by b0noI** [comment]
please use this commit for reproduction: https://github.com/GeminiAgentsToolkit/gemini-agents-toolkit/commit/c2abef95b78097c3f9e969b38fa8549d9a0a4771

in the more latest version I have remediated this with the hack (prompt that explains to the backend how NOT to call functions :) )

**Comment by Ark-kun** [comment]
Hello. I will forward your issue to the model quality team. The SDK cannot do anything here since the model response and the erro comes from the backend.

```
Finish message: Malformed function call:
print(157.3 * 1.04)
```

As you see, to solve your question, the model generated the following "function call": `print(157.3 * 1.04)`, but it wasn't calling any function, just doing multiplication.

To reduce cases like this you can try using tool_config to force function calling. When using forced function calling there are additional constraint decoding features enabled that can reduce/eliminate the probability of invalid function call. https://github.com/googleapis/python-aiplatform/pull/3534

**Comment by RomanBrodetski** [comment]
Also consistently reproduces for me with `gemini-1.5-pro`.

Potentially relevant tool that I'm passing:
```
    {
        name: "solc",
        description: "Compiles a solidity contract. Returns ABI if successful, and error message if not",
        parameters:
            {
                type: FunctionDeclarationSchemaType.OBJECT,
                properties: {
                    code: {
                        type: FunctionDeclarationSchemaType.STRING,
                        description: "The solidity code to compile"
                    }
                }
            }
    },
 ```
 
Thanks!

**Comment by mrothroc** [comment]
I am seeing this with gemini-2.0-flash.  It seems like you cannot have both function calling and a return type on the same call.  It would be nice if the documentation explained that because I spent a lot of time trying to figure this out.

**Comment by nclmasiero** [comment]
> I am seeing this with gemini-2.0-flash. It seems like you cannot have both function calling and a return type on the same call. It would be nice if the documentation explained that because I spent a lot of time trying to figure this out.

Can you elaborate? I'm not sure I know what a return type is in this context. Does the Gemini API requires you to specify what return type to expect in the tool definition?

For context:
I have an AI agent with multiple functions to call. I keep getting this error and have no way to debug nor understand what's the root cause. I've tried to:
- rewrite the prompt more clearly
- pinpoint the function, if just one, that causes the error (without success)
- format functions in different ways (e.g. specifying full array of elements in a list instead of a function to add/remove elements)

So this is recurrent to me and I don't know what else to try.

**Comment by roei-shlezinger** [comment]
My most effective solution was to break down the functions into numerous small, clear functions, each unambiguously defined. I had attempted to create complex functions, assuming that if they didn't work, they would simply return no result. However, they actually caused errors. To improve the situation, I ensured that each function received only one mandatory variable. Additionally, I aimed to have the instructions for each function defined solely within the function itself, while preventing any overlap in instructions between different functions or with the instructions in the system prompt.

**Comment by mrothroc** [comment]
> Can you elaborate? I'm not sure I know what a return type is in this context. Does the Gemini API requires you to specify what return type to expect in the tool definition?

I'm sorry, I should have said "response type".  This setting allows you to force the LLM to respond with a specific type, e.g. application/json.  If you set this, then you can't have function calling.

**Comment by viperfx** [comment]
I see this happening with complex tools. Gemini just cannot handle it, where other models can - even older OpenAI or Anthropic ones can handle it fine. 

Gemini should do a better job of clearly displaying what the issue is. Providing a finish reason with MALFORMED_FUNCTION_CALL is just not enough. 

If there is an issue with the tools spec, we need to understand what the issue is so it can be fixed.

**Comment by kwbtdisk** [comment]
I was also struggling with this issue but discovered the following solution after trying various approaches.

Regarding the root cause, I suspect there might be a parser bug/defect in the Google infrastructure where Gemini responses are being parsed.

### "MALFORMED_FUNCTION_CALL" Problem with Gemini Function Calls and How to Avoid It

- Reproduction conditions:
  - Prompt tokens 120k >=
  - Attempting to output large content through tool calls
    - Issue: Only function arguments containing extensive content get truncated
    - Additionally, if the argument is specified as `required` in the function definition's JSON schema, a `MALFORMED_FUNCTION_CALL` Error occurs immediately after Gemini's response, as it's considered an invalid function call

- Workaround: Always use text responses
  - When making requests:
    - Set `function_calling_config` to `NONE` in the tools (function calls) configuration
    - Set up tools (functions) definitions
    - Pass them as normal system or user messages
  - When handling responses:
    - Gemini will attempt to use the tools correctly, but the responses will be returned in text format (Python code)
    - Parse the function name and argument contents
    - Example:
      ```markdown
      ```tool_code
      print(default_api.file_editor_create(purpose = "somevalue", path = "")
      ```

### Example of Parsing tool_code (typescript)

```ts
/**
 * For Gemini response, Parses a tool_code string format and extracts the function name and arguments
 * @param {string} inputString - The string enclosed in ```tool_code ... ```
 * @returns {Object} An object containing the functionName and args
 */
export function parseGeminiToolResponse(inputString: string): {
  functionName: string
  args: Record<string, any>
} {
  // Step 1: Extract the code content, removing the ```tool_code and ``` markers
  const codeContent = inputString
    .trim()
    .replace(/^```tool_code\n/, '')
    .replace(/\n```$/, '')

  // Step 2: Extract the function name
  const functionNameMatch = codeContent.match(/print\(default_api\.([a-zA-Z_]+)\(/)
  const functionName = functionNameMatch ? functionNameMatch[1] : null

  if (!functionName) {
    throw new Error('Could not extract function name')
  }

  // Step 3: Extract the argument string and prepare for JSON parsing
  let argsString = codeContent
    .replace(/print\(default_api\.[a-zA-Z_]+\(/, '')
    .trim()
    .replace(/\)+$/, '')
    .trim()

  // Convert the argument format to valid JSON format
  // First, handle the first argument which doesn't have a leading comma
  argsString = argsString.replace(/^([a-zA-Z_]+)\s*=\s*/, '"$1": ')

  // Then handle all subsequent arguments
  argsString = argsString.replace(/,\s*([a-zA-Z_]+)\s*=\s*/g, ', "$1": ')

  // Ensure proper escaping of backslashes in the JSON string
  // argsString = argsString.replace(/\\{2,}/g, (match) => '\\'.repeat(match.length / 2))
  argsString = argsString.replace(/\\\'/g, "'")
  // close the string if it's not closed
  if (argsString.endsWith('"') === false) {
    argsString = argsString + '"'
  }
  // Wrap in curly braces to make it a valid JSON object
  const jsonString = `{${argsString}}`

  // Step 4: Parse the JSON string to get the arguments object
  let args
  try {
    args = new Function(
      [
        'const True = true', // Python code convertion
        'const False = false', // Python code convertion
        'const None = null', // Python code convertion
        'const Infinity = 1/0', // Python code convertion
        'const NaN = 0/0', // Python code convertion
        'return ' + jsonString,
      ].join('\n'),
    )()
  } catch (error) {
    console.error('Error parsing JSON:', error)
    console.error('JSON string:', jsonString)
    throw new Error('Failed to parse arguments as JSON')
  }

  return {
    functionName,
    args,
  }
}

// Example usage:
// const inputString = "```tool_code\nprint(default_api.file_editor_create(file_text = \"content\", path = \"./path\", purpose = \"purpose\"))\n```";
// const result = parseToolCode(inputString);
// console.log(result);
```

**Comment by lozhn** [comment]
I may have found something: I tried to switch to None function calls and parse generated code manually, Claude generated code for this like so
```python

def extract_code_block(text: str) -> str:
    """Extract code from ```tool_code``` or ```python``` blocks."""
    # Match either ```tool_code or ```python followed by any content until ```
    pattern = r"```(?:tool_code|python)\n(.*?)```"
    match = re.search(pattern, text, re.DOTALL)
    if match:
        return match.group(1).strip()
    return text
```

Which is working fine for simple cases

<img width="442" alt="Image" src="https://github.com/user-attachments/assets/e48852ec-5a45-4416-b165-b6ba0983bc79" />

But fails if you want Markdown with code

<img width="421" alt="Image" src="https://github.com/user-attachments/assets/c4a8780d-28b8-43c2-8dbf-9973300234ba" />

Why it may be related? Because gemini-2.0-flash fails a lot when I use function calling with Mode.ANY in the same exact manner

```
ValueError: Cannot get the Candidate text.
Response candidate content has no parts (and thus no text). The candidate is likely blocked by the safety filters.
Content:
{
  "role": "model"
}
Candidate:
{
  "content": {
    "role": "model"
  },
  "finish_reason": "MALFORMED_FUNCTION_CALL",
  "finish_message": "Malformed function call: print(default_api.chat_response(message=\""
}
```

WDYT @b0noI ?

**Comment by matthew29tang** [comment]
The following Generative AI modules in the Vertex AI SDK are deprecated as of June 24, 2025 and will be removed on June 24, 2026: vertexai.generative_models, vertexai.language_models, vertexai.vision_models, vertexai.tuning, vertexai.caching. Please use the [Google Gen AI SDK](https://pypi.org/project/google-genai/) to access these features. See [the migration guide](https://cloud.google.com/vertex-ai/generative-ai/docs/deprecations/genai-vertexai-sdk) for details. You can continue using all other Vertex AI SDK modules, as they are the recommended way to use the API.

**Comment by yGuy** [comment]
This seems unrelated to the SDK as I see the same bad behavior with raw API (curl/node fetch) access. 
So I think this is still an issue, but it's not in the SDK but rather a bug in the API endpoints.

@matthew29tang could you elaborate on your last comment, please? Are you saying this is fixed, or does not apply to the new SDK?

**Comment by matthew29tang** [comment]
Since this SDK is deprecated for generative models (Ex. Gemini), we aren't actively monitoring here for these kinds of issues.

If you are still encountering this issue on the latest models (2.5), please open a new issue in our new SDK https://github.com/googleapis/python-genai and feel free to reference this thread.

**Comment by agshinrajabov** [comment]
Still seeing this in production with Gemini 2.0 Flash and 2.5 Flash. The root cause appears to be in Gemini's output tokenizer — it generates unescaped strings inside function call JSON arguments, especially with large text content.

Three workarounds that actually work:

Add "Ensure all string values in function call arguments are properly JSON-escaped" to your system prompt — biggest single fix (~90% reduction).
Retry on MALFORMED with temperature=0 — deterministic decoding produces valid JSON more reliably.
Handle the empty content field explicitly — the SDK throws when you call response.text() on a MALFORMED response.
We open-sourced our production fix as a TypeScript library: [gemini-heal](https://github.com/emotixco/gemini-heal) — adaptive rate limiting, MALFORMED recovery, and circuit breaking. The core logic (prompt-level fix + retry strategy) applies to any language.
